src/processor/uri_parse.cpp

```diff
@@ -0,0 +1,162 @@
+            query_remaining = query_remaining.substr(separator_pos + 1);
+        } else {
+            parameter = query_remaining;
+            query_remaining = {};
+        }
+
+        std::string_view key;
+        ddwaf_object value;
+
+        // Check if it's in the key=value format
+        //  - key= is considered an empty string value
+        //  - key is considered a flag
+        auto assignment_pos = parameter.find_first_of('=');
+        if (assignment_pos == std::string_view::npos) {
+            key = parameter;
+
+            // Ignore empty keys
+            if (key.empty()) {
+                continue;
+            }
+
+            // We can consider this a modifier rather than a kv pair
+            ddwaf_object_bool(&value, true);
+        } else {
+            key = parameter.substr(0, assignment_pos);
+            // Ignore empty keys
+            if (key.empty()) {
+                continue;
+            }
+
+            auto value_str = parameter.substr(assignment_pos + 1);
+            ddwaf_object_stringl(&value, value_str.data(), value_str.size());
+        }
+
+        // Check if the key has the array suffix ([]) and strip it, if the suffix
+        // contains an index ([index]), we just consider it a separate key
+        if (key.ends_with("[]")) {
+            key.remove_suffix(sizeof("[]") - 1);
+        }
+
+        auto it = query_keys.find(key);
+        if (it == query_keys.end()) {
+            query_keys.emplace(key, value);
+        } else {
+            // Duplicate! We need to create an array or add to it
+            if (it->second.type != DDWAF_OBJ_ARRAY) {
+                ddwaf_object array;
+                ddwaf_object_array(&array);
+                ddwaf_object_array_add(&array, &it->second);
+                it->second = array;
+            }
+
+            ddwaf_object_array_add(&it->second, &value);
+        }
+    }
+
+    ddwaf_object query;
+    ddwaf_object_map(&query);
+    for (auto &[key, value] : query_keys) {
+        ddwaf_object_map_addl(&query, key.data(), key.size(), &value);
+    }
+
+    return query;
+}
+
+} // namespace
+
+// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+std::pair<ddwaf_object, object_store::attribute> uri_parse_processor::eval_impl(
+    const unary_argument<std::string_view> &input, processor_cache & /*cache*/,
+    ddwaf::timer & /*deadline*/) const
+{
+    const object_store::attribute attr =
+        input.ephemeral ? object_store::attribute::ephemeral : object_store::attribute::none;
+
+    auto decomposed = uri_parse(input.value);
+    if (!decomposed.has_value()) {
+        return {};
+    }
+
+    ddwaf_object scheme = string_view_to_object(decomposed->scheme);
+    ddwaf_object userinfo = string_view_to_object(decomposed->authority.userinfo);
+    ddwaf_object host = string_view_to_object(decomposed->authority.host);
+
+    ddwaf_object port;
+    if (auto [res, value] = from_string<uint16_t>(decomposed->authority.port); res) {
```

Comment:
you're calling `from_string` already when parsing, though you discard the value; in any case, this can never fail if the string is non-empty. maybe you could add the port to decomposed in number form already.